Build fails with AppSync Gradle plugin version 2.6.18

The build is failing during the `afterEvaluate` phase.

```
. . .
09:44:27.476 [DEBUG] [org.gradle.internal.progress.DefaultBuildOperationExecutor] Completing Build operation 'Configure project :app'
09:44:27.476 [null] [org.gradle.internal.progress.DefaultBuildOperationExecutor]
09:44:27.476 [DEBUG] [org.gradle.internal.progress.DefaultBuildOperationExecutor] Completing Build operation 'Configure build'
09:44:27.478 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]
09:44:27.478 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] FAILURE: Build failed with an exception.
09:44:27.478 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]
09:44:27.478 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] * Where:
09:44:27.478 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] Build file '/Users/*******/******/appsync/AppSyncExperiment2/app/build.gradle' line: 2
09:44:27.478 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]
09:44:27.479 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] * What went wrong:
09:44:27.479 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] A problem occurred evaluating project ':app'.
09:44:27.479 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] > com/moowork/gradle/node/NodePlugin
09:44:27.479 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]
09:44:27.479 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] * Try:
09:44:27.479 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]  Run with --scan to get full insights.
09:44:27.479 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]
09:44:27.479 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] * Exception is:
09:44:27.480 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] org.gradle.api.GradleScriptException: A problem occurred evaluating project ':app'.
09:44:27.480 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]   at org.gradle.groovy.scripts.internal.DefaultScriptRunnerFactory$ScriptRunnerImpl.run(DefaultScriptRunnerFactory.java:92)
09:44:27.480 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]   at org.gradle.configuration.DefaultScriptPluginFactory$ScriptPluginImpl$2.run(DefaultScriptPluginFactory.java:199)
09:44:27.480 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]   at org.gradle.configuration.ProjectScriptTarget.addConfiguration(ProjectScriptTarget.java:77)
09:44:27.480 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]   at org.gradle.configuration.DefaultScriptPluginFactory$ScriptPluginImpl.apply(DefaultScriptPluginFactory.java:204)
. . .
```

I think the recent change which made `NodePlugin` as `compileOnly` dependency in gradle is what causing the issue. https://github.com/awslabs/aws-mobile-appsync-sdk-android/blob/a57a80a7f31085115bb1b7479d88dc3d89f994d3/aws-android-sdk-appsync-gradle-plugin/build.gradle#L15

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,17 @@
 # Change Log - AWS AppSync SDK for Android
 
+## [Release 2.6.19](https://github.com/awslabs/aws-mobile-appsync-sdk-android/releases/tag/release_v2.6.19)
+
+### Enhancements
+
+* Improve dependency injection for gradle plugin.
+Uses `implementation` instead of `compile` when using higher than gradle version 2.
+Note: gradle version 0 and 1 are note supported.
+
+### Bug Fixes
+
+* Fixed dependency `com.moowork.gradle:gradle-node-plugin:1.0.0` in gradle plugin
+
 ## [Release 2.6.18](https://github.com/awslabs/aws-mobile-appsync-sdk-android/releases/tag/release_v2.6.18)
 
 ### New Features

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@
 * Improve dependency injection for gradle plugin.
 Uses `implementation` instead of `compile` when using higher than gradle version 2.
 Note: gradle version 0 and 1 are note supported.
+* Improve `BasicCognitoUserPoolsAuthProvider` retrieval of token.
 
 ### Bug Fixes
 

--- a/README.md
+++ b/README.md
@@ -20,7 +20,7 @@ In the project's `build.gradle`, add the following dependency in
 the build script:
 
 ```groovy
-    classpath 'com.amazonaws:aws-android-sdk-appsync-gradle-plugin:2.6.18'
+    classpath 'com.amazonaws:aws-android-sdk-appsync-gradle-plugin:2.6.19'
 ```
 
 **Sample project's build.gradle**
@@ -31,7 +31,7 @@ the build script:
         // ..other code..
         dependencies {
             classpath 'com.android.tools.build:gradle:3.0.1'
-            classpath 'com.amazonaws:aws-android-sdk-appsync-gradle-plugin:2.6.18'
+            classpath 'com.amazonaws:aws-android-sdk-appsync-gradle-plugin:2.6.19'
             // NOTE: Do not place your application dependencies here; they belong
             // in the individual module build.gradle files
         }
@@ -49,7 +49,7 @@ In the app's `build.gradle`, add the following plugin:
 Add the following dependency:
 
 ```groovy
-    compile 'com.amazonaws:aws-android-sdk-appsync:2.6.18'
+    compile 'com.amazonaws:aws-android-sdk-appsync:2.6.19'
 ```
 
 **Sample app's build.gradle**
@@ -62,7 +62,7 @@ Add the following dependency:
     }
     dependencies {
         // Typical dependencies
-        compile 'com.amazonaws:aws-android-sdk-appsync:2.6.18'
+        compile 'com.amazonaws:aws-android-sdk-appsync:2.6.19'
     }
 ```
 

--- a/aws-android-sdk-appsync-compiler/src/generated/kotlin/com/apollographql/android/Version.kt
+++ b/aws-android-sdk-appsync-compiler/src/generated/kotlin/com/apollographql/android/Version.kt
@@ -1,3 +1,3 @@
 // Generated file. Do not edit!
 package com.apollographql.android
-val VERSION = "2.6.18"
+val VERSION = "2.6.19"

--- a/aws-android-sdk-appsync-gradle-plugin/build.gradle
+++ b/aws-android-sdk-appsync-gradle-plugin/build.gradle
@@ -12,7 +12,7 @@ dependencies {
 
     compileOnly gradleApi()
     compileOnly localGroovy()
-    compileOnly "com.moowork.gradle:gradle-node-plugin:1.0.0"
+    compile "com.moowork.gradle:gradle-node-plugin:1.0.0" // compileOnly
     compile 'com.android.tools.build:gradle:2.3.1' // compileOnly TODO Check if this can be upgraded.
     compile 'com.squareup.moshi:moshi:1.5.0' // impl
 

--- a/aws-android-sdk-appsync-gradle-plugin/src/main/groovy/com/apollographql/apollo/gradle/ApolloPlugin.groovy
+++ b/aws-android-sdk-appsync-gradle-plugin/src/main/groovy/com/apollographql/apollo/gradle/ApolloPlugin.groovy
@@ -40,157 +40,164 @@ import org.gradle.api.tasks.compile.JavaCompile
 import javax.inject.Inject
 
 class ApolloPlugin implements Plugin<Project> {
-  private static final String NODE_VERSION = "6.7.0"
-  public static final String TASK_GROUP = "apollo"
-  private static final String APOLLO_DEP_GROUP = "com.amazonaws"
-  private static final String RUNTIME_DEP_NAME = "aws-android-sdk-appsync-runtime"
+    private static final String NODE_VERSION = "6.7.0"
+    public static final String TASK_GROUP = "apollo"
+    private static final String AMAZON_DEP_GROUP = "com.amazonaws"
+    private static final String RUNTIME_DEP_NAME = "aws-android-sdk-appsync-runtime"
 
-  private Project project
-  private final FileResolver fileResolver
+    private Project project
+    private final FileResolver fileResolver
 
-  @Inject
-  public ApolloPlugin(FileResolver fileResolver) {
-    this.fileResolver = fileResolver
-  }
-
-  @Override void apply(Project project) {
-    this.project = project
-    if (project.plugins.hasPlugin(AppPlugin) || project.plugins.hasPlugin(LibraryPlugin) || project.plugins.hasPlugin(
-        JavaPlugin)) {
-      applyApolloPlugin()
-    } else {
-      throw new IllegalArgumentException(
-          "Apollo plugin couldn't be applied. The Android or Java plugin must be configured first")
+    @Inject
+    public ApolloPlugin(FileResolver fileResolver) {
+        this.fileResolver = fileResolver
     }
-  }
 
-  private void applyApolloPlugin() {
-    setupNode()
-    project.extensions.create(ApolloExtension.NAME, ApolloExtension)
-    createSourceSetExtensions()
-
-    def compileDepSet = project.configurations.getByName("compile").dependencies
-    project.getGradle().addListener(new DependencyResolutionListener() {
-      @Override
-      void beforeResolve(ResolvableDependencies resolvableDependencies) {
-        def apolloRuntimeDep = compileDepSet.find { dep ->
-          dep.group == APOLLO_DEP_GROUP
-          dep.name == RUNTIME_DEP_NAME
+    @Override
+    void apply(Project project) {
+        this.project = project
+        if (project.plugins.hasPlugin(AppPlugin) || project.plugins.hasPlugin(LibraryPlugin) || project.plugins.hasPlugin(
+                JavaPlugin)) {
+            applyApolloPlugin()
+        } else {
+            throw new IllegalArgumentException(
+                    "Apollo plugin couldn't be applied. The Android or Java plugin must be configured first")
         }
-        if (apolloRuntimeDep != null && apolloRuntimeDep.version != VersionKt.VERSION && !apolloRuntimeDep instanceof ProjectDependency) {
-          throw new GradleException(
-              "apollo-runtime version ${apolloRuntimeDep.version} isn't compatible with the apollo-gradle-plugin version ${VersionKt.VERSION}")
+    }
+
+    private void applyApolloPlugin() {
+        setupNode()
+        project.extensions.create(ApolloExtension.NAME, ApolloExtension)
+        createSourceSetExtensions()
+
+        def gradleVersion = project.getGradle().getGradleVersion()
+        def depType = "implementation"
+        if (gradleVersion.startsWith("2")) {
+            depType = "compile"
         }
-        if (System.getProperty("apollographql.skipRuntimeDep") != "true" && apolloRuntimeDep == null) {
-          compileDepSet.add(project.dependencies.create("$APOLLO_DEP_GROUP:$RUNTIME_DEP_NAME:$VersionKt.VERSION"))
+        def depSet = project.configurations.getByName(depType).dependencies
+
+        project.getGradle().addListener(new DependencyResolutionListener() {
+            @Override
+            void beforeResolve(ResolvableDependencies resolvableDependencies) {
+                def appsyncRuntimeDep = depSet.find { dep ->
+                    dep.group == AMAZON_DEP_GROUP
+                    dep.name == RUNTIME_DEP_NAME
+                }
+                if (appsyncRuntimeDep != null && appsyncRuntimeDep.version != VersionKt.VERSION && !appsyncRuntimeDep instanceof ProjectDependency) {
+                    throw new GradleException(
+                            "apollo-runtime version ${appsyncRuntimeDep.version} isn't compatible with the apollo-gradle-plugin version ${VersionKt.VERSION}")
+                }
+                if (System.getProperty("apollographql.skipRuntimeDep") != "true" && appsyncRuntimeDep == null) {
+                    depSet.add(project.dependencies.create("$AMAZON_DEP_GROUP:$RUNTIME_DEP_NAME:$VersionKt.VERSION"))
+                }
+            }
+
+            @Override
+            void afterResolve(ResolvableDependencies resolvableDependencies) {
+                project.getGradle().removeListener(this)
+            }
+        })
+
+        project.tasks.create(ApolloCodeGenInstallTask.NAME, ApolloCodeGenInstallTask.class)
+        addApolloTasks()
+    }
+
+    private void addApolloTasks() {
+        Task apolloIRGenTask = project.task("generateApolloIR")
+        apolloIRGenTask.group(TASK_GROUP)
+        Task apolloClassGenTask = project.task("generateApolloClasses")
+        apolloClassGenTask.group(TASK_GROUP)
+
+        if (isAndroidProject()) {
+            getVariants().all { v ->
+                addVariantTasks(v, apolloIRGenTask, apolloClassGenTask, v.sourceSets)
+            }
+            project.android.testVariants.each { tv ->
+                addVariantTasks(tv, apolloIRGenTask, apolloClassGenTask, tv.sourceSets)
+            }
+        } else {
+            getSourceSets().all { sourceSet ->
+                addSourceSetTasks(sourceSet, apolloIRGenTask, apolloClassGenTask)
+            }
         }
-      }
-
-      @Override
-      void afterResolve(ResolvableDependencies resolvableDependencies) {
-        project.getGradle().removeListener(this)
-      }
-    })
-
-    project.tasks.create(ApolloCodeGenInstallTask.NAME, ApolloCodeGenInstallTask.class)
-    addApolloTasks()
-  }
-
-  private void addApolloTasks() {
-    Task apolloIRGenTask = project.task("generateApolloIR")
-    apolloIRGenTask.group(TASK_GROUP)
-    Task apolloClassGenTask = project.task("generateApolloClasses")
-    apolloClassGenTask.group(TASK_GROUP)
-
-    if (isAndroidProject()) {
-      getVariants().all { v ->
-        addVariantTasks(v, apolloIRGenTask, apolloClassGenTask, v.sourceSets)
-      }
-      project.android.testVariants.each { tv ->
-        addVariantTasks(tv, apolloIRGenTask, apolloClassGenTask, tv.sourceSets)
-      }
-    } else {
-      getSourceSets().all { sourceSet ->
-        addSourceSetTasks(sourceSet, apolloIRGenTask, apolloClassGenTask)
-      }
-    }
-  }
-
-  private void addVariantTasks(Object variant, Task apolloIRGenTask, Task apolloClassGenTask, Collection<?> sourceSets) {
-    ApolloIRGenTask variantIRTask = createApolloIRGenTask(variant.name, sourceSets)
-    ApolloClassGenTask variantClassTask = createApolloClassGenTask(variant.name)
-    variant.registerJavaGeneratingTask(variantClassTask, variantClassTask.outputDir)
-    apolloIRGenTask.dependsOn(variantIRTask)
-    apolloClassGenTask.dependsOn(variantClassTask)
-  }
-
-  private void addSourceSetTasks(SourceSet sourceSet, Task apolloIRGenTask, Task apolloClassGenTask) {
-    String taskName = "main".equals(sourceSet.name) ? "" : sourceSet.name
-
-    ApolloIRGenTask sourceSetIRTask = createApolloIRGenTask(sourceSet.name, [sourceSet])
-    ApolloClassGenTask sourceSetClassTask = createApolloClassGenTask(sourceSet.name)
-    apolloIRGenTask.dependsOn(sourceSetIRTask)
-    apolloClassGenTask.dependsOn(sourceSetClassTask)
-
-    JavaCompile compileTask = (JavaCompile) project.tasks.getByName("compile${taskName.capitalize()}Java")
-    compileTask.source += project.fileTree(sourceSetClassTask.outputDir)
-    compileTask.dependsOn(apolloClassGenTask)
-  }
-
-  private void setupNode() {
-    project.plugins.apply NodePlugin
-    NodeExtension nodeConfig = project.extensions.findByName("node") as NodeExtension
-    nodeConfig.download = true
-    nodeConfig.version = NODE_VERSION
-  }
-
-  private ApolloIRGenTask createApolloIRGenTask(String sourceSetOrVariantName, Collection<Object> sourceSets) {
-    String taskName = String.format(ApolloIRGenTask.NAME, sourceSetOrVariantName.capitalize())
-    ApolloIRGenTask task = project.tasks.create(taskName, ApolloIRGenTask) {
-      group = TASK_GROUP
-      description = "Generate an IR file using apollo-codegen for ${sourceSetOrVariantName.capitalize()} GraphQL queries"
-      dependsOn(ApolloCodeGenInstallTask.NAME)
-      sourceSets.each { sourceSet ->
-        inputs.file(sourceSet.graphql).skipWhenEmpty()
-      }
     }
 
-    ImmutableList.Builder<String> sourceSetNamesList = ImmutableList.builder();
-    sourceSets.each { sourceSet -> sourceSetNamesList.add(sourceSet.name) }
-    task.init(sourceSetOrVariantName, sourceSetNamesList.build(), project.apollo)
-    return task
-  }
-
-  private ApolloClassGenTask createApolloClassGenTask(String name) {
-    String taskName = String.format(ApolloClassGenTask.NAME, name.capitalize())
-    ApolloClassGenTask task = project.tasks.create(taskName, ApolloClassGenTask) {
-      group = TASK_GROUP
-      description = "Generate Android classes for ${name.capitalize()} GraphQL queries"
-      dependsOn(getProject().getTasks().findByName(String.format(ApolloIRGenTask.NAME, name.capitalize())));
-      source = project.tasks.findByName(String.format(ApolloIRGenTask.NAME, name.capitalize())).outputFolder
-      include "**${File.separatorChar}*API.json"
+    private void addVariantTasks(Object variant, Task apolloIRGenTask, Task apolloClassGenTask, Collection<?> sourceSets) {
+        ApolloIRGenTask variantIRTask = createApolloIRGenTask(variant.name, sourceSets)
+        ApolloClassGenTask variantClassTask = createApolloClassGenTask(variant.name)
+        variant.registerJavaGeneratingTask(variantClassTask, variantClassTask.outputDir)
+        apolloIRGenTask.dependsOn(variantIRTask)
+        apolloClassGenTask.dependsOn(variantClassTask)
     }
-    task.init(name, project.apollo)
-    return task
-  }
 
-  private void createSourceSetExtensions() {
-    getSourceSets().all { sourceSet ->
-      sourceSet.extensions.create(GraphQLSourceDirectorySet.NAME, GraphQLSourceDirectorySet, sourceSet.name,
-          fileResolver)
+    private void addSourceSetTasks(SourceSet sourceSet, Task apolloIRGenTask, Task apolloClassGenTask) {
+        String taskName = "main".equals(sourceSet.name) ? "" : sourceSet.name
+
+        ApolloIRGenTask sourceSetIRTask = createApolloIRGenTask(sourceSet.name, [sourceSet])
+        ApolloClassGenTask sourceSetClassTask = createApolloClassGenTask(sourceSet.name)
+        apolloIRGenTask.dependsOn(sourceSetIRTask)
+        apolloClassGenTask.dependsOn(sourceSetClassTask)
+
+        JavaCompile compileTask = (JavaCompile) project.tasks.getByName("compile${taskName.capitalize()}Java")
+        compileTask.source += project.fileTree(sourceSetClassTask.outputDir)
+        compileTask.dependsOn(apolloClassGenTask)
     }
-  }
 
-  private boolean isAndroidProject() {
-    return project.hasProperty('android') && project.android.sourceSets
-  }
+    private void setupNode() {
+        project.plugins.apply NodePlugin
+        NodeExtension nodeConfig = project.extensions.findByName("node") as NodeExtension
+        nodeConfig.download = true
+        nodeConfig.version = NODE_VERSION
+    }
 
-  private Object getSourceSets() {
-    return (isAndroidProject() ? project.android.sourceSets : project.sourceSets)
-  }
+    private ApolloIRGenTask createApolloIRGenTask(String sourceSetOrVariantName, Collection<Object> sourceSets) {
+        String taskName = String.format(ApolloIRGenTask.NAME, sourceSetOrVariantName.capitalize())
+        ApolloIRGenTask task = project.tasks.create(taskName, ApolloIRGenTask) {
+            group = TASK_GROUP
+            description = "Generate an IR file using apollo-codegen for ${sourceSetOrVariantName.capitalize()} GraphQL queries"
+            dependsOn(ApolloCodeGenInstallTask.NAME)
+            sourceSets.each { sourceSet ->
+                inputs.file(sourceSet.graphql).skipWhenEmpty()
+            }
+        }
 
-  private DomainObjectCollection<BaseVariant> getVariants() {
-    return project.android.hasProperty(
-        'libraryVariants') ? project.android.libraryVariants : project.android.applicationVariants
-  }
+        ImmutableList.Builder<String> sourceSetNamesList = ImmutableList.builder();
+        sourceSets.each { sourceSet -> sourceSetNamesList.add(sourceSet.name) }
+        task.init(sourceSetOrVariantName, sourceSetNamesList.build(), project.apollo)
+        return task
+    }
+
+    private ApolloClassGenTask createApolloClassGenTask(String name) {
+        String taskName = String.format(ApolloClassGenTask.NAME, name.capitalize())
+        ApolloClassGenTask task = project.tasks.create(taskName, ApolloClassGenTask) {
+            group = TASK_GROUP
+            description = "Generate Android classes for ${name.capitalize()} GraphQL queries"
+            dependsOn(getProject().getTasks().findByName(String.format(ApolloIRGenTask.NAME, name.capitalize())));
+            source = project.tasks.findByName(String.format(ApolloIRGenTask.NAME, name.capitalize())).outputFolder
+            include "**${File.separatorChar}*API.json"
+        }
+        task.init(name, project.apollo)
+        return task
+    }
+
+    private void createSourceSetExtensions() {
+        getSourceSets().all { sourceSet ->
+            sourceSet.extensions.create(GraphQLSourceDirectorySet.NAME, GraphQLSourceDirectorySet, sourceSet.name,
+                    fileResolver)
+        }
+    }
+
+    private boolean isAndroidProject() {
+        return project.hasProperty('android') && project.android.sourceSets
+    }
+
+    private Object getSourceSets() {
+        return (isAndroidProject() ? project.android.sourceSets : project.sourceSets)
+    }
+
+    private DomainObjectCollection<BaseVariant> getVariants() {
+        return project.android.hasProperty(
+                'libraryVariants') ? project.android.libraryVariants : project.android.applicationVariants
+    }
 }

--- a/aws-android-sdk-appsync/build.gradle
+++ b/aws-android-sdk-appsync/build.gradle
@@ -30,9 +30,9 @@ dependencies {
     compile 'com.google.code.findbugs:jsr305:3.0.1' // api
     compile "com.squareup.okhttp3:okhttp:3.8.1" // impl
 
-    compile 'com.amazonaws:aws-android-sdk-core:2.6.18' // api
-    compileOnly 'com.amazonaws:aws-android-sdk-cognitoidentityprovider:2.6.18'
-    compileOnly 'com.amazonaws:aws-android-sdk-s3:2.6.18'
+    compile "com.amazonaws:aws-android-sdk-core:$aws_version" // api
+    compileOnly "com.amazonaws:aws-android-sdk-cognitoidentityprovider:$aws_version"
+    compileOnly "com.amazonaws:aws-android-sdk-s3:$aws_version"
 
     compileOnly 'org.eclipse.paho:org.eclipse.paho.client.mqttv3:1.1.0'
     compileOnly 'org.eclipse.paho:org.eclipse.paho.android.service:1.1.1'

--- a/aws-android-sdk-appsync/src/main/java/com/amazonaws/mobileconnectors/appsync/sigv4/BasicCognitoUserPoolsAuthProvider.java
+++ b/aws-android-sdk-appsync/src/main/java/com/amazonaws/mobileconnectors/appsync/sigv4/BasicCognitoUserPoolsAuthProvider.java
@@ -17,6 +17,7 @@
 
 package com.amazonaws.mobileconnectors.appsync.sigv4;
 
+import com.amazonaws.mobileconnectors.appsync.AWSAppSyncClient;
 import com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoDevice;
 import com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoUserPool;
 import com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoUserSession;
@@ -25,10 +26,17 @@ import com.amazonaws.mobileconnectors.cognitoidentityprovider.continuations.Chal
 import com.amazonaws.mobileconnectors.cognitoidentityprovider.continuations.MultiFactorAuthenticationContinuation;
 import com.amazonaws.mobileconnectors.cognitoidentityprovider.handlers.AuthenticationHandler;
 
+import java.util.concurrent.Semaphore;
+
+/**
+ * Basic retrieval of the Cognito Userpools token. The user must have already signed-in before using
+ * this class as the retrieval mechanism in the {@link AWSAppSyncClient#builder()}
+ */
 public class BasicCognitoUserPoolsAuthProvider implements CognitoUserPoolsAuthProvider {
 
     private CognitoUserPool userPool;
     private String token;
+    private String lastTokenRetrievalFailureMessage;
 
     public BasicCognitoUserPoolsAuthProvider(CognitoUserPool userPool) {
         this.userPool = userPool;
@@ -38,33 +46,50 @@ public class BasicCognitoUserPoolsAuthProvider implements CognitoUserPoolsAuthPr
     /**
      * Fetches token from the Cognito User Pools client for the current user.
      */
-    private void fetchToken() {
+    private synchronized void fetchToken() {
+        final Semaphore semaphore = new Semaphore(0);
+        lastTokenRetrievalFailureMessage = null;
         userPool.getCurrentUser().getSessionInBackground(new AuthenticationHandler() {
             @Override
             public void onSuccess(CognitoUserSession userSession, CognitoDevice newDevice) {
                 token = userSession.getAccessToken().getJWTToken();
+                semaphore.release();
             }
 
             @Override
             public void getAuthenticationDetails(AuthenticationContinuation authenticationContinuation, String userId) {
-
+                lastTokenRetrievalFailureMessage = "Cognito Userpools is not signed-in";
+                semaphore.release();
             }
 
             @Override
             public void getMFACode(MultiFactorAuthenticationContinuation continuation) {
-
+                lastTokenRetrievalFailureMessage = "Cognito Userpools is not signed-in";
+                semaphore.release();
             }
 
             @Override
             public void authenticationChallenge(ChallengeContinuation continuation) {
-
+                lastTokenRetrievalFailureMessage = "Cognito Userpools is not signed-in";
+                semaphore.release();
             }
 
             @Override
             public void onFailure(Exception exception) {
-
+                lastTokenRetrievalFailureMessage = "Cognito Userpools failed to get session";
+                semaphore.release();
             }
         });
+
+        try {
+            semaphore.acquire();
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Interrupted waiting for Cognito Userpools token.", e);
+        }
+
+        if (lastTokenRetrievalFailureMessage != null) {
+            throw new RuntimeException(lastTokenRetrievalFailureMessage);
+        }
     }
 
     @Override

--- a/build.gradle
+++ b/build.gradle
@@ -3,6 +3,7 @@
 
 buildscript {
     ext.kotlin_version = '1.2.30'
+    ext.aws_version = VERSION_NAME
     repositories {
         google()
         jcenter()

--- a/gradle.properties
+++ b/gradle.properties
@@ -13,7 +13,7 @@ org.gradle.jvmargs=-Xmx1536m
 org.gradle.parallel=true
 
 GROUP=com.amazonaws
-VERSION_NAME=2.6.18
+VERSION_NAME=2.6.19
 
 POM_URL=https://github.com/awslabs/aws-mobile-appsync-sdk-android
 POM_SCM_URL=https://github.com/awslabs/aws-mobile-appsync-sdk-android

